ansible_base/models/resource.py

```diff
@@ -0,0 +1,132 @@
+            s_id = service_id().split('-')[0]
+        except ProgrammingError:
+            s_id = "00000000"
+
+        return (
+            super()
+            .get_queryset()
+            .annotate(
+                _computed_service_id=models.Case(
+                    models.When(_service_id__isnull=False, then=models.F('_service_id')), default=models.Value(s_id), output_field=models.CharField()
+                ),
+                _ansible_id=Concat("_computed_service_id", models.Value(":"), models.F("resource_id"), output_field=models.CharField()),
+            )
+        )
+
+
+class ResourceType(models.Model):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.resource_registry = get_registry()
+
+    content_type = models.OneToOneField(ContentType, on_delete=models.CASCADE, related_name="resource_type", unique=True)
+    externally_managed = models.BooleanField()
+    migrated = models.BooleanField(null=False, default=False)
+    resource_type = models.CharField(max_length=256, unique=True, db_index=True)
+
+    @property
+    def serializer_class(self):
+        return self.get_resource_config()["managed_serializer"]
+
+    def get_resource_config(self):
+        return self.resource_registry.get_config_for_model(model=self.content_type.model_class())
+
+
+class Resource(models.Model):
+    objects = ResourceManager()
+
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE, related_name="resources")
+
+    # this has to accommodate integer and UUID object IDs
+    object_id = models.TextField(null=False)
+    content_object = GenericForeignKey('content_type', 'object_id')
+
+    _service_id = models.CharField(null=True, max_length=8)
+
+    # we're not using this as the primary key because the ansible_id can change if the object is
+    # externally managed.
+    resource_id = models.UUIDField(default=uuid.uuid4, db_index=True, unique=True)
+
+    # human readable name for the resource
+    name = models.CharField(max_length=256, null=True)
+
+    @property
+    def service_id(self):
+        return self._computed_service_id
+
+    @property
+    def ansible_id(self):
+        return self._ansible_id
+
+    @ansible_id.setter
+    def ansible_id(self, val):
+        s_id, r_id = val.split(":")
+        if not service_id().startswith(s_id):
+            self._service_id = s_id
+
+        self.resource_id = r_id
+
+    @property
+    def resource_type(self):
+        return self.content_type.resource_type.resource_type
+
+    class Meta:
+        unique_together = ('content_type', 'object_id')
+        indexes = [
+            models.Index(fields=["content_type", "object_id"]),
+        ]
+
+
+class PostgresTransaction(models.Model):
+    transaction = models.BigIntegerField(primary_key=True)
+    gid = models.CharField(max_length=200)
+    prepared = models.DateTimeField()
+
+    class Meta:
+        managed = False
+        db_table = "pg_prepared_xacts"
```

Comment:
Something that can be done later. We should limit the surface area of this model/api endpoint. The set of transactions here should be limited to the ones that were created by our service/app. Thinking about users/customers with shared database servers. This might be a non-issue and Postgres users/roles might be "the way".